internal/proxy/origins/prometheus/matrix.go

```diff
@@ -137,23 +167,38 @@ func (me *MatrixEnvelope) CropToSize(sz int, t time.Time, lur timeseries.Extent)
 
 	for _, s := range me.Data.Result {
 		tmp := s.Values[:0]
+		wg := sync.WaitGroup{}
+		mtx := sync.Mutex{}
 		for _, r := range s.Values {
-			t = r.Timestamp.Time()
-			if _, ok := removals[t]; !ok {
-				tmp = append(tmp, r)
-			}
+			wg.Add(1)
+			go func(p model.SamplePair) {
+				mtx.Lock()
+				t = p.Timestamp.Time()
```

Comment:
Each of the goroutines per `me.Data.Result` can be updating `t` concurrently because the lock is specific per `me.Data.Result`; you are not guaranteed to get the correct value where you want it.

Reply:
👍